QueueManager initial commit

milljm opened the pull request:
QueueManager is properly launching and checking jobs using PBS

Changes to some tests were necessary to handle relative path
issues involved with directory creation.

Refs #9237

milljm commented on python/TestHarness/schedulers/Job.py:
```diff
@@ -53,6 +54,26 @@ def getDAG(self):
+        if self.__dag_clone == None:
+            self.__dag_clone = self.__dag.clone()
+        return self.__dag_clone
+
```
_really_ wish I can make this somehow 'automatic' when the class gets instanced. But the desired DAG does not get created until later.

Is there a way to do what I am asking with out an external call to setOriginalDAG ?

milljm commented on python/TestHarness/schedulers/Scheduler.py:
```diff
@@ -292,16 +292,14 @@ def schedule(self, testers):
-                tester.setStatus('QUEUED', tester.bucket_pending)
+        # Set original DAG state now, before we launch any jobs
+        for job_container in job_dag.topological_sort():
+            job_container.setOriginalDAG()
```
With my first comment above, this is what I am talking about. Having to set the original DAG now that its complete. It must be done for each job object because it is _not_ shared.

permcody replied:
You have a few options. We are working on the assumption that you build the DAG, and then you use the DAG. We don't build the DAG, use it a little, then add to it some more. If we assume that workflow, then you definitely can remove the explicit call to "set the DAG" (not sure I love that name). You simply set the DAG the very first time you try to use it. To do that, you'd have to modify the underlying DAG class so that when you call topological sort or whatever else on it, that you check some Boolean and call that set method on it before you start any other operation. I'll go through your PR and see if I understand what you are doing.

milljm replied:
hmmm, perhaps when we perform a  dag.ind_nodes() do get a list of concurrent jobs... when thats called, we should assume the dag is ready to rock...

milljm commented on python/TestHarness/util.py:
```diff
@@ -517,7 +517,10 @@ def readOutput(f, options, max_size=100000):
+    if options.no_trimmed_output or options.sep_files == True:
+        output += f.read()
+
+    else:
```
--no-trimmed-output as requested

Commit:
QueueManager initial commit

QueueManager is properly launching and checking jobs using PBS

Refs #9237

moosebuild commented:
Job [Documentation](https://www.moosebuild.org/job/114678/) on 7da1d91 wanted to post the following:

View the site [here](http://mooseframework.org/docs/PRs/9819/site)

Commit:
Test Modifications

Changes to the following tests were necessary in order
to allow them to pass while using PBS.

Refs #9237

permcody commented on python/contrib/dag/__init__.py:
```diff
@@ -146,6 +146,22 @@ def delete_downstreams(self, node, graph=None):
         return deleted_nodes
 
+    # Added by the MOOSE group
+    def reverse_edges(self, graph=None):
```
You might want a better name. Maybe `cloned_reversed`? From the name alone, it sounds like it might work on the current graph, not return a new graph.

permcody commented on python/TestHarness/schedulers/RunPBS.py:
```diff
@@ -0,0 +1,286 @@
+            json_session = self.getData(test_unique, job_id=True)
+            util.runCommand('qrls %s' % (json_session['job_id']))
+
+    def populateTemplateParams(self, job_container, template):
```
This method is OK here for now, but the expansion process is actually pretty neutral. I could be factored out into the QueueManager class and should be once we add a new queueing Scheduler.

permcody commented on python/TestHarness/schedulers/RunPBS.py:
```diff
@@ -0,0 +1,286 @@
+
+        return template
+
+    def prepareQueueScript(self, template):
```
This method is not PBS specific, could be moved into the QueueManager class.

permcody commented on python/TestHarness/schedulers/RunPBS.py:
```diff
@@ -0,0 +1,286 @@
+        f.close()
+
+    # Process the tester output and run processResults
+    def testOutput(self, job_container):
```
Looks like this could be moved to QueueManager.

permcody commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -0,0 +1,403 @@
+            queueJobs(run_jobs=[list of jobs])
+
+           Print status:
+            queueJobs(status_jobs=[list of jobs])
```
Honestly, these really should just be separate methods. Since `queueJobs` is your entry point, this method would just turn around and call these two methods as appropriate. No need to combine these functions into a single method with a giant "if" statement.

permcody commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -0,0 +1,403 @@
+                if test_unique not in self.__session_data.keys():
+                    continue
+
+                # Yo dawg, I heard you like DAGs
```
This isn't the right usage for this meme. "Yo dawg" is all about nesting things inside of things. You just have a lot at one level!

permcody commented on python/TestHarness/schedulers/QueueManager.py:
```diff
@@ -0,0 +1,403 @@
+                # run these new jobs
+                self.queueJobs(run_jobs=next_job_group)
+
+    def checkStatus(self, job_container):
```
This whole method could go away if you just put the extra dependencies right inside of the "expanded DAG" . In fact, it would just fall back to the statusing method in the Scheduler base class! You think you're obsessed with DAGs?! I'm the one that really wants one DAG to rule them all. Bahahaha

permcody reviewed (commented):
This is looking pretty good. I think this design could really benefit from using a single expanded DAG during the status step. Right now you have to check the status (a huge subroutine) on each node to know whether or not you can process the reverse DAG.

Commit:
Refactor methods in RunPBS

Address comments, optimize a few things.
This is a working commit, but there is a lot of things to do. Namely
we need to the Scheduler's threading model.

Refs #9237

jasondhales commented:
I know you guys are working on this, and we appreciate it.  As it stands now, we don't have support for running our assessment cases, and so the sooner we can do that for end-of-FY reporting, the better.

permcody commented:
I'm testing the current implementation now. I hope to have this in today. We have a lot of work to do, but this will get you up and running again, then we can continue to improve the capability and speed of this new work.

permcody commented:
I'm reasonably convinced that this addition does not negatively affect the current functionality of the test harness. Also, my initial test of the new PBS functionality looks good aside from several things I've talked with Jason about. This should be good to go into the framework.
